### DISCUSSION TITLE
Allow just the "parser" to be overridden (instead of the whole Evaluator).

### DISCUSSION THREAD

**Discussion by bennadel** [alpinejs/alpine#4094] [3. Feature ideas]
This is just a random thought for a fun experiment; and I don't understand all the implications. Take this with a grain of salt.

As a fun experiment, I took the Angular.js 1.x parser and used to create a custom Evaluator that is CSP-compatible but still allows for _much_ (not all) of the normal expressions in Alpine.js:

https://www.bennadel.com/blog/4612-using-the-angular-js-parser-to-comply-with-csp-in-alpine-js-3-13-5.htm

However, my evaluator isn't really compatible with Alpine.js because it doesn't do things like inject Magics. Looking at the CSP-build, I see that you're calling internal methods like `injectMagics()` and `tryCatch` that are not exposed on the public `Alpine` object.

If there was a way in which all I had to do was override the `generateEvaluatorFromString()`, then I think it would be easier to replace the parser without having to completely override the evaluator.

Anyway, I'm just thinking out loud, and tested the edges of the framework. Happy Friday!

<img src="https://github.com/alpinejs/alpine/assets/563690/d7f06bbf-86e8-4fb5-a58d-ae9eb825afbd" width="400" />

**Comment by SimoTod** [comment]
I think the setEvaluator function was added at the beginning to allow that customisation needed for the CSP build but overtime some private logic crept into the evaluator itself. I believe the original intention was to use it more as a private api (hence it's not massively documented). 

You can in theory add alpine as a dependency for your project, import those functions (i know, you won't get any bug fixing but those low level functions hardly change) and build your own evaluator.

**Reply by bennadel** [reply]
That makes sense as I've seen some Issues/tickets around CSP-support only being added recently. I suppose, at the least, I can just look at the stuff that is implemented privately and then re-create it a custom evaluator. I think the Magics injection is the primary issue in the current CSP evaluator.

Thanks for the feedback :)

**Reply by SimoTod** [reply]
No worries. You can also try to create a PR with a setEvaluatorParser that just replace that function (exposing it externally) and try to pitch it to the maintainer.

It's not a lot of work and you can use your dev brannch to test your new parser.

To be honest, it makes more sense to me to replace the parser than the full evaluator.

**Comment by ekwoka** [comment]
I don't think the evaluator injects the magics.

The data already has the magics.

You just need to handle the context being provided.

**Reply by bennadel** [reply]
@ekwoka I'm only basing it on this line:

https://github.com/alpinejs/alpine/blob/main/packages/csp/src/evaluator.js#L22

But, to be honest, I really haven't explored what Magics are yet. On my list of things to explore.

**Reply by ekwoka** [reply]
Ah, hmm, right. Using `Alpine.$data` would only get the magics as bound to the root, not the element.

This should be refactored to use a context passed in to the evaluators...

**Reply by bennadel** [reply]
All things in perspective. It doesn't seem like there's all that much demand for CSP (at least not according to my Googling). I happen to have a fairly strict CSP on my blog, which is why I am curious about it.

**Reply by ekwoka** [reply]
Yes, it's mostly been a "if you want it, come try to make it work" kind of situation.
